### ISSUE TITLE
Repeated Duplicate errors in the CheckpointVerifier

### ISSUE THREAD

**Issue by teor2345** [ZcashFoundation/zebra#1259]
**Version**

`zebrad` main branch, 2020-11-05

**Platform**

`Linux ... 5.4.72 #1-NixOS SMP Sat Oct 17 08:11:24 UTC 2020 x86_64 GNU/Linux`

**Description**

On testnet, I see repeated `Duplicate` errors from the `CheckpointVerifier`. The block locator remains at the same height - the state does not make any progress. The number of queued blocks in the download set remains close to 2000.

Restarting `zebrad` resolves this issue, so it's a bug in the block download, verify, and commit pipeline. (Rather than a bug in the cached state on disk.)

I tried this:

Running `zebrad` on testnet with a few local peers in addition to the peers from the seeder. My local network connection is slower and higher latency than the typical `zebrad` developer's network connection. Testnet also has fewer and slower peers than mainnet.

I expected to see this happen:

`zebrad` syncs to the Sapling checkpoint (or the testnet best tip).

Instead, this happened:

`zebrad` stalls with repeated `Duplicate` errors.

**Context**

The `Duplicate` errors don't contain enough information to find the source of this bug. It could be an issue with the syncer, download set, checkpoint verifier, or state service. PR #1260 improves the `CheckpointVerifier` errors, and renames `Duplicate` to `AlreadyVerified`.

Some of the `CheckpointVerifier` tests are disabled, I'll submit a PR that re-enables them.

**Comment by teor2345** [comment]
PR #1260 improves the `CheckpointVerifier` errors, and renames `Duplicate` to `AlreadyVerified`.

**Comment by teor2345** [comment]
A restart resolves this issue.

**Comment by teor2345** [comment]
Using the error improvements from #1260 and `zebrad-log-filter` (with some fields removed), I see the following logs. Similar logs repeat indefinitely.

Note that the block locator is stuck near 218400, even though checkpoint 218800 was verified. After some time, I saw a block locator at 218479, even though the checkpoints for this range are 218400 and 218800.

It looks like the checkpoint verifier is verifying the blocks, but they don't get committed to the state, or they take a very long time to get committed to the state.

```
Nov 06 14:19:49.399  INFO sync:extend_tips:download_and_verify{hash=block::Hash("                                                                                                    high: 217820                                                                                                                                                                                   
hash: 000480c37632bd561434cb1ce4c20f87afe6d8840418df586cc9e7fb381e4662                                                                                                                         
")}:checkpoint: zebra_consensus::checkpoint: verified checkpoint range block_count=400 current_range=(Excluded(Height(217600)), Included(Height(218000)))                                      
Nov 06 14:19:57.455  INFO sync: zebrad::components::sync: waiting for pending blocks tips.len=1 in_flight=2371 lookahead_limit=2000                                                            
Nov 06 14:19:57.456  INFO sync: zebrad::components::sync: extending tips tips.len=1 in_flight=2000 lookahead_limit=2000                                                                        
Nov 06 14:19:57.972  INFO sync:extend_tips:download_and_verify{hash=block::Hash("                                                                                                              
high: 218249                                                                                                                                                                                   
hash: 00040627c48b7913dc25080267c16e84623b01d36f948487781f4a9f7c23aee4                                                                                                                         
")}:checkpoint: zebra_consensus::checkpoint: verified checkpoint range block_count=400 current_range=(Excluded(Height(218000)), Included(Height(218400)))
Nov 06 14:20:03.484  INFO sync: zebrad::components::sync: exhausted prospective tip set
Nov 06 14:20:03.484  INFO sync: zebrad::components::sync: waiting to restart sync timeout=45s
Nov 06 14:20:48.487  INFO sync: zebrad::components::sync: starting sync, obtaining new tips
Nov 06 14:20:48.487  INFO sync:obtain_tips: zebra_state::util: created block locator tip_height=Height(218400) min_locator_height=218301 locators=[Height(218400), Height(218399), Height(21839
8), Height(218396), Height(218392), Height(218384), Height(218368), Height(218336), Height(218301)]
Nov 06 14:21:02.472  INFO sync: zebrad::components::sync: extending tips tips.len=1 in_flight=499 lookahead_limit=2000
Nov 06 14:21:02.992  INFO sync:obtain_tips:download_and_verify{hash=block::Hash("
high: 218789
hash: 000b6ec9593bcc9deb8a235f78297197fea34c87c5773d34d66eace3923140d6
")}:checkpoint: zebra_consensus::checkpoint: verified checkpoint range block_count=400 current_range=(Excluded(Height(218400)), Included(Height(218800)))
Nov 06 14:21:02.995  INFO sync:obtain_tips:download_and_verify{hash=block::Hash("
high: 218617
hash: 000c4f95e00e0ce6e395dece2a71f3945fd27530c5d7743d6b2183ea3aef293a
")}:checkpoint: zebra_consensus::checkpoint: e=Err(AlreadyVerified { height: Height(218617), verified_height: Height(218800) })
Nov 06 14:21:03.038  INFO sync:obtain_tips:download_and_verify{hash=block::Hash("
high: 218584
hash: 0005232723298964e548a2bd5773ffa9da20af1700382342b1ddb913a0461754
")}:checkpoint: zebra_consensus::checkpoint: e=Err(AlreadyVerified { height: Height(218584), verified_height: Height(218800) })
Nov 06 14:21:03.073  INFO sync:obtain_tips:download_and_verify{hash=block::Hash("
high: 218549
hash: 000c319abec7267b8bffd1e769462e5fabd99fc7f3f6ad78046ec3fd752ef22d
")}:checkpoint: zebra_consensus::checkpoint: e=Err(AlreadyVerified { height: Height(218549), verified_height: Height(218800) })
Nov 06 14:21:03.156  INFO sync:obtain_tips:download_and_verify{hash=block::Hash("
high: 218480
hash: 00022cf2154400be54e142370062e669c2b0a5196e9bbd3d3084270812f51961
")}:checkpoint: zebra_consensus::checkpoint: e=Err(AlreadyVerified { height: Height(218480), verified_height: Height(218800) })
Nov 06 14:21:03.235  INFO sync:obtain_tips:download_and_verify{hash=block::Hash("
high: 218683
hash: 0006788a1679f385323be909411b96c5dc0a250e4a4fdd4ae4214bd0d1ef3ebc
")}:checkpoint: zebra_consensus::checkpoint: e=Err(AlreadyVerified { height: Height(218683), verified_height: Height(218800) })
Nov 06 14:21:11.404  WARN sync: zebrad::components::sync: e=Checkpoint(AlreadyVerified { height: Height(218480), verified_height: Height(218800) })
Nov 06 14:21:11.405  INFO sync: zebrad::components::sync: waiting to restart sync timeout=45s
```

**Comment by teor2345** [comment]
The relevant metrics are:

Progress
* checkpoint_queued_max_height: 219396
* checkpoint_queued_continuous_height: 218998
* checkpoint_verified_height: 218800
* checkpoint_processing_next_height: 218800
* state_committed_block_height: 218479

Queue
* checkpoint_queued_slots: 570
* queued_block_rate [1s]: peaks of up to 60 blocks per second, separated by stalls that last for minutes
* state_queued_block_count: 316
  * this seems bad - the checkpoint verifier should only verify and commit contiguous intervals

**Comment by teor2345** [comment]
Network health:
* 14 peers, 9-14 peers are ready most of the time, occasional peaks when all peers are not ready
* failed candidate peers grows by about 30 every 5 minutes

**Comment by teor2345** [comment]
This error also happens in CI, for example:
https://github.com/ZcashFoundation/zebra/pull/1262/checks?check_run_id=1362438343#step:8:398

**Comment by teor2345** [comment]
This error also happens on mainnet.

Here are the relevant metrics, using PRs #1261 and #1262:

Heights:
* state_finalized_commited_block_height = 106134
* state_finalized_queued_max_height = checkpoint_verified_height = 106604
* checkpoint_queued_continuous_height = 106645
* checkpoint_queued_max_height = 107231

Queue Lengths:
* checkpoint_queued_slots = 554
* state_finalized_queued_block_count = 319

It seems like some blocks are getting stalled or dropped between the verifier and the state. (Or the verifier is skipping some blocks during verification or commit.)

**Comment by teor2345** [comment]
I'm using this grafana dashboard:
https://gist.github.com/teor2345/f627b7e0288a9313d212481c8af5c947

(It probably has too many metrics on it - I'll delete some when I know which ones are useful.)

**Comment by teor2345** [comment]
Fixed by #1321 and #1322.
